iaso/api/fhir/location.py

```diff
@@ -0,0 +1,304 @@
+
+    def get_queryset(self):
+        """
+        Get OrgUnit queryset filtered for the current user
+        """
+        return (
+            OrgUnit.objects.filter_for_user(self.request.user)
+            .select_related("org_unit_type", "parent", "version__data_source")
+            .prefetch_related("groups")
+        )
+
+    def create_fhir_bundle(self, locations_data: List[Dict], total_count: int, request) -> Dict[str, Any]:
+        """
+        Create a FHIR Bundle response containing Location resources
+        """
+        entries = []
+        for location_data in locations_data:
+            entries.append(
+                {
+                    "resource": location_data,
+                    "fullUrl": f"{request.build_absolute_uri().split('?')[0]}/{location_data['id']}",
+                }
+            )
+
+        bundle = {
+            "resourceType": "Bundle",
+            "id": "search-results",
+            "meta": {"lastUpdated": "2024-01-01T00:00:00Z"},
+            "type": "searchset",
+            "total": total_count,
+            "entry": entries,
+        }
+
+        # Add pagination links
+        links = self._build_pagination_links(request, total_count)
+        if links:
+            bundle["link"] = links
+
+        return bundle
+
+    def _get_pagination_params(self, request):
+        """Get pagination parameters from request"""
+        try:
+            page_size = int(request.query_params.get("_count", 20))
+            page_size = min(page_size, 100)  # Max 100
+        except (TypeError, ValueError):
+            page_size = 20
+
+        try:
+            offset = int(request.query_params.get("_skip", 0))
+        except (TypeError, ValueError):
+            offset = 0
+
+        return page_size, offset
+
+    def _build_pagination_links(self, request, total_count: int) -> List[Dict[str, str]]:
+        """Build FHIR pagination links"""
+        links = []
+
+        # Get pagination parameters
+        page_size, offset = self._get_pagination_params(request)
+
+        # Self link
+        self_url = request.build_absolute_uri().split("?")[0]
+        if request.query_params:
+            self_url += f"?{request.query_params.urlencode()}"
+        links.append({"relation": "self", "url": self_url})
+
+        # Next link
+        if offset + page_size < total_count:
+            next_params = request.query_params.copy()
+            next_params["_skip"] = offset + page_size
+            next_url = f"{request.build_absolute_uri().split('?')[0]}?{next_params.urlencode()}"
+            links.append({"relation": "next", "url": next_url})
+
+        # Previous link
+        if offset > 0:
+            prev_params = request.query_params.copy()
+            prev_params["_skip"] = max(0, offset - page_size)
+            prev_url = f"{request.build_absolute_uri().split('?')[0]}?{prev_params.urlencode()}"
+            links.append({"relation": "previous", "url": prev_url})
+
+        return links
```

Comment:
This whole pagination logic could have been implemented as a `LimitOffsetPagination` class:

```python
class FHIRPaginator(pagination.LimitOffsetPagination):
    default_limit = 20
    limit_query_param = "_count"
    limit_query_description = 'Number of results to return per page.'
    offset_query_param = "_skip"
    offset_query_description = "The initial index from which to return the results."
    max_limit = 100

    …

```

Avoiding mixing "views" and "pagination" logic in the same place could provide some clarity.

References:

- https://www.django-rest-framework.org/api-guide/pagination/#limitoffsetpagination
- https://github.com/encode/django-rest-framework/blob/e454758fb6edf1dcf5aa5417a388b940c871469c/rest_framework/pagination.py#L373